routing/payment_lifecycle.go

```diff
@@ -66,158 +67,243 @@ type shardOutcome struct {
+
+			return [32]byte{}, nil, fmt.Errorf("No successful " +
+				"attempts nor failure found!")
+		}
+
+		remValue := p.totalAmount - sentAmt
+
+	ShardLoop:
+		// While we still have amount to send, and we didn't aleady
+		// encounter a success or parmanent failure, send another
+		// shard.
+		for remValue > 0 && !(success || failure != nil) {
 			// Before we attempt this next payment, we'll check to see if either
 			// we've gone past the payment attempt timeout, or the router is
 			// exiting. In either case, we'll stop this payment attempt short. If a
 			// timeout is not applicable, timeoutChan will be nil.
 			select {
 			case <-p.timeoutChan:
-				// Mark the payment as failed because of the
-				// timeout.
-				err := p.router.cfg.Control.Fail(
-					p.paymentHash, channeldb.FailureReasonTimeout,
-				)
-				if err != nil {
-					return [32]byte{}, nil, err
-				}
-
 				errStr := fmt.Sprintf("payment attempt not completed " +
 					"before timeout")
 
-				return [32]byte{}, nil, newErr(ErrPaymentAttemptTimeout, errStr)
-
-			case <-p.router.quit:
-				// The payment will be resumed from the current state
-				// after restart.
-				return [32]byte{}, nil, ErrRouterShuttingDown
+				terminalFailure = &paymentFailure{
+					failureCode: channeldb.FailureReasonTimeout,
+					err:         newErr(ErrPaymentAttemptTimeout, errStr),
+				}
+				break ShardLoop
 
+			// Fall through if we haven't hit our time limit.
 			default:
-				// Fall through if we haven't hit our time limit, or
-				// are expiring.
 			}
 
 			// Create a new payment attempt from the given payment session.
 			rt, err := p.paySession.RequestRoute(
-				p.totalAmount, uint32(p.currentHeight),
+				remValue, uint32(p.currentHeight),
 			)
 			if err != nil {
 				log.Warnf("Failed to find route for payment %x: %v",
 					p.paymentHash, err)
 
 				// Convert error to payment-level failure.
-				failure := errorToPaymentFailure(err)
-
-				// If we're unable to successfully make a payment using
-				// any of the routes we've found, then mark the payment
-				// as permanently failed.
-				saveErr := p.router.cfg.Control.Fail(
-					p.paymentHash, failure,
-				)
-				if saveErr != nil {
-					return [32]byte{}, nil, saveErr
+				routeFailure = &paymentFailure{
+					failureCode: errorToPaymentFailure(err),
+					err:         err,
 				}
 
 				// If there was an error already recorded for this
 				// payment, we'll return that.
 				if lastError != nil {
-					return [32]byte{}, nil, errNoRoute{lastError: lastError}
+					routeFailure.err = errNoRoute{lastError: lastError}
 				}
 
-				// Terminal state, return.
-				return [32]byte{}, nil, err
+				break ShardLoop
 			}
 
-			// With the route in hand, laumnch a new shard.
-			attempt, outcome, err = shardHandler.launch(rt)
-			if err != nil {
-				return [32]byte{}, nil, err
-			}
+			// Launch a shard along the route, getting the outcome
+			// delivered on the channel.
+			shardHandler.launchAndCollect(
```

Comment:
The field in the routing payload itself? I had assumed that would be set only at the very start since they all should share the same value.

Reply:
That's a good question. Since it is a feature on the `route` itself, I imagined it would be set by the creator of the route (the `PaymentSession`), and that is in the follow-up: https://github.com/lightningnetwork/lnd/pull/3967/commits/d0ebe3d4c3f5574167a950bbe87be4fbe8df7c29#diff-8b9a8672616d7a3dd85c4dcbc216ebdfR169

It is noted as a "hack" tho, do you think it shouldn't be the PaymentSessions' responsibility @joostjager ?

Reply:
I think it shouldn't be. In this pr, the payment session doesn't need to be aware of the bigger scheme of the payment at all. It just receives the amount remaining and finds a route for that. I think the payment lifecycle should do it. Maybe even lift the call to `newRoute` out of the payment session.

Reply:
The nice this about doing it in the payment session is that it has everything it needs from the `LightningPayment` to make that decision. I think it also needs to know from being able to determine if the target node supports MPP?

The payment lifecycle is indifferent, it just launches the routes the paymentsession serves it.